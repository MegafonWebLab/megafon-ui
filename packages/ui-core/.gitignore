.idea
*.iml
node_modules

src/**/*.css
src/index.ts
src/components/**/*.d.ts
dist
coverage
.docz
docs
.vscode
.DS_Store
