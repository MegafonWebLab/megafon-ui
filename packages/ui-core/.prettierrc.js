module.exports = {
    ...require('@megafon/frontend-presets/prettier'),
};
