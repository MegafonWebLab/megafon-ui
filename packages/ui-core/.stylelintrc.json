{
  "rules": {
    "color-hex-case": "upper",
    "color-no-invalid-hex": true,
    "font-family-name-quotes": "always-where-recommended",
    "function-calc-no-unspaced-operator": true,
    "function-comma-space-after": "always",
    "function-comma-space-before": "never",
    "number-leading-zero": "always",
    "number-no-trailing-zeros": true,
    "string-no-newline": true,
    "string-quotes": "single",
    "length-zero-no-unit": true,
    "unit-case": "lower",
    "unit-no-unknown": true,
    "value-keyword-case": "lower",
    "value-no-vendor-prefix": true,
    "property-case": "lower",
    "property-no-unknown": true,
    "keyframe-declaration-no-important": true,
    "declaration-bang-space-after": "never",
    "declaration-block-no-duplicate-properties": true,
    "declaration-block-trailing-semicolon": "always",
    "declaration-block-semicolon-space-before": "never",
    "block-no-empty": true,
    "block-opening-brace-newline-after": "always",
    "block-closing-brace-newline-before": "always",
    "selector-attribute-brackets-space-inside": "never",
    "max-nesting-depth": 5,
    "indentation": 4,
    "declaration-bang-space-before": "always"
  }
}
