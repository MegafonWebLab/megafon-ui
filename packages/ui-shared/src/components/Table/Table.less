@import '~@megafon/ui-core/styles/base.less';

@b: e('.mfui-beta-table');

@{b} {
    @firstColumnWidth: 144px;
    @gradientSize: 48px;
    @shadowSize: 6px;
    @transparent: rgba(255, 255, 255, 0);

    .fontFamily();
    .commonFont();
    display: flex;

    &__scroll {
        display: grid;
        grid-template-columns: auto auto auto;
        max-width: 100%;
        overflow: auto;
        position: relative;
        border: 1px solid @spbSky2;

        @media @desktop {
            max-height: 500px;
        }
    }

    &__top-shadow {
        position: sticky;
        top: 0;
        width: 100%;
        display: block;
        grid-column-end: 4;
        grid-column-start: 1;
        z-index: 2;
        margin-bottom: -@shadowSize;
        height: @shadowSize;
        background: linear-gradient(359.94deg, rgba(0, 0, 0, 0.0001) 0.06%, rgba(0, 0, 0, 0.03) 38.08%, rgba(0, 0, 0, 0.08) 74.98%);
    }

    &__left-shadow {
        position: sticky;
        z-index: 2;
        width: @shadowSize;
        left: 0;
        top: 0;
        bottom: 0;
        margin-right: -@shadowSize;
        background: linear-gradient(270deg, rgba(0, 0, 0, 0.0001) 16.3%, rgba(0, 0, 0, 0.03) 51.81%, rgba(0, 0, 0, 0.08) 86.27%);
    }

    &__bottom-gradient {
        display: block;
        height: @gradientSize;
        width: 100%;
        margin-top: -@gradientSize;
        position: sticky;
        z-index: 1;
        bottom: 0;
        background: linear-gradient(to top, @clearWhite 25%, @transparent);
        grid-column-end: 4;
        grid-column-start: 1;
    }

    &__right-gradient {
        display: block;
        height: 100%;
        width: @gradientSize;
        position: sticky;
        z-index: 1;
        right: 0;
        margin-left: -@gradientSize;
        background: linear-gradient(to left, @clearWhite 25%, @transparent);
    }

    &__table {
        border-spacing: 0;
        border-width: 0;
        border-collapse: separate;
        background-color: @clearWhite;
    }

    &__cell {
        padding: 24px;
        box-sizing: border-box;
        background-color: @clearWhite;

        @media @mobileSM {
            padding-left: 16px;
            padding-right: 16px;
        }
    }

    &__cell:first-child {
        left: 0;
        min-width: @firstColumnWidth;
        width: @firstColumnWidth;
        position: relative;
        border-right: 1px solid @spbSky2;
    }

    &__cell:not(:first-child) {
        max-width: 288px;
    }

    &__row_head &__cell {
        .small();
        height: 72px;
        background-color: @spbSky0;
        color: fade(@freshAsphalt, 50%);
        padding-bottom: 12px;
        vertical-align: bottom;
    }

    &__row:not(:last-child) &__cell {
        border-bottom: 1px solid @spbSky2;
    }

    // mods

    &_min-cell-size_small {
        @{b}__cell:not(:first-child) {
            min-width: 72px;
        }
    }

    &_min-cell-size_large {
        @{b}__cell:not(:first-child) {
            min-width: 144px;
        }
    }

    &_fix-column {
        @{b}__cell:first-child {
            position: sticky;
        }

        @{b}__left-shadow {
            left: @firstColumnWidth;
        }

        @{b}__bottom-gradient {
            z-index: 3;
        }
    }

    &:not(&_touch) {
        ::-webkit-scrollbar {
            width: 8px;
            height: 8px;
        }

        ::-webkit-scrollbar-track {
            background-color: @spbSky1;
        }

        ::-webkit-scrollbar-thumb {
            border-radius: 4px;
            background-color: @spbSky2;
        }

        ::-webkit-scrollbar-corner {
            background-color: @clearWhite;
        }

        @{b}__scroll {
            scrollbar-color: @spbSky2 @spbSky1;
            scrollbar-width: thin;
        }
    }
}
