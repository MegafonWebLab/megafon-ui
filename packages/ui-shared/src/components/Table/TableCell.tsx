import * as React from 'react';
import * as PropTypes from 'prop-types';

export interface ITableCell {
    children?: React.ReactNode;
}

const TableCell: React.FC<ITableCell> = ({ children }) => <>{children && children}</>;

TableCell.propTypes = {
    children: PropTypes.node,
};

export default TableCell;
