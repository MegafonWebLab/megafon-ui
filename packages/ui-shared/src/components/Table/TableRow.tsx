import * as React from 'react';
import * as PropTypes from 'prop-types';
import { ITableCell } from './TableCell';

export interface ITableRow {
    /** Ряд таблицы как заголовок */
    head?: boolean;
    children: Array<React.ReactElement<ITableCell>>;
}

const TableRow: React.FC<ITableRow> = ({ children }) => <>{children}</>;

TableRow.propTypes = {
    head: PropTypes.bool,
    children: PropTypes.arrayOf(PropTypes.element.isRequired).isRequired,
};

export default TableRow;
