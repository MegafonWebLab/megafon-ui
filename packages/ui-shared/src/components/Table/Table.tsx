import * as React from 'react';
import * as PropTypes from 'prop-types';
import throttle from 'lodash.throttle';
import convert from 'htmr';
import { cnCreate, detectTouch } from '@megafon/ui-core';
import './Table.less';
import { ITableRow } from './TableRow';

export interface ITable {
    /** Дополнительный класс корневого элемента */
    className?: string;
    /** Фиксация первой колонки */
    fixColumn?: boolean;
    /** Минимальный размер ячеек */
    minCellSize?: 'small' | 'large';
    children: Array<React.ReactElement<ITableRow>>;
}

const cn = cnCreate('mfui-beta-table');
const Table: React.FC<ITable> = ({
    className,
    fixColumn = true,
    minCellSize = 'large',
    children,
}) => {
    const scrollRef = React.useRef<HTMLDivElement | null>(null);

    const [isTopShadow, setTopShadow] = React.useState(false);
    const [isLeftShadow, setLeftShadow] = React.useState(false);
    const [isRightGradient, setRightGradient] = React.useState(false);
    const [isBottomGradient, setBottomGradient] = React.useState(false);
    const [isTouchDevice, setTouchDevice] = React.useState(false);

    const handleTableScroll = React.useCallback(() => {
        const { current: scrollNode } = scrollRef;

        if (!scrollNode) {
            return;
        }

        const {
            scrollHeight,
            clientHeight,
            scrollTop,
            scrollWidth,
            clientWidth,
            scrollLeft,
        } = scrollNode;

        const deltaHeight = scrollHeight - clientHeight;
        const deltaWidth = scrollWidth - clientWidth;

        setRightGradient(deltaWidth !== Math.floor(scrollLeft));
        setBottomGradient(deltaHeight !== Math.floor(scrollTop));
        setTopShadow(!!scrollTop);
        setLeftShadow(!!scrollLeft);
    }, []);

    React.useEffect(() => {
        const handleResize = throttle(handleTableScroll, 300);

        handleTableScroll();
        window.addEventListener('resize', handleResize);

        return () => {
            window.removeEventListener('resize', handleResize);
        };
    }, [handleTableScroll]);

    React.useEffect(() => {
        setTouchDevice(detectTouch());
    }, []);

    const renderRows = () =>
        React.Children.map(children, ({ props: { children: cells, head } }) => (
            <tr className={cn('row', { head })}>
                {React.Children.map(cells, ({ props: { children: cell } }) => (
                    <td className={cn('cell')}>
                        {typeof cell === 'string' ? convert(cell) : cell}
                    </td>
                ))}
            </tr>
        ));

    return (
        <div
            className={cn(
                {
                    'fix-column': fixColumn,
                    'min-cell-size': minCellSize,
                    touch: isTouchDevice,
                },
                [className]
            )}
        >
            <div
                className={cn('scroll')}
                ref={scrollRef}
                onScroll={handleTableScroll}
            >
                {isTopShadow && <div className={cn('top-shadow')} />}
                {isLeftShadow && <div className={cn('left-shadow')} />}
                <table className={cn('table')}>
                    <tbody>{renderRows()}</tbody>
                </table>
                {isRightGradient && <div className={cn('right-gradient')} />}
                {isBottomGradient && <div className={cn('bottom-gradient')} />}
            </div>
        </div>
    );
};

Table.propTypes = {
    className: PropTypes.string,
    fixColumn: PropTypes.bool,
    minCellSize: PropTypes.oneOf(['small', 'large']),
    children: PropTypes.arrayOf(PropTypes.element.isRequired).isRequired,
};

export default Table;
