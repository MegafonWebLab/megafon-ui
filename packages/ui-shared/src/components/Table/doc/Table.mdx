---
name: Table
menu: Конструктор
route: components/Table
---

import { Props } from 'docz';
import Table from '../Table';
import TableRow from '../TableRow';
import Example from './Table.example.mdx'

# Table

<CopyToClipboardBox text="import { Table, TableRow, TableCell } from '@megafon/ui-shared';" />

<DoczTabs>
    <Example />
    <div>
        <Props of={Table} />
        <h2>TableRow</h2>
        <Props of={TableRow} />
    </div>
</DoczTabs>
