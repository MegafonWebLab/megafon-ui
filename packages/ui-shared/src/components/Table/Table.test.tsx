import * as React from 'react';
import { mount, shallow } from 'enzyme';
import { cnCreate } from '@megafon/ui-core';
import Table, { ITable } from './Table';
import TableRow from './TableRow';
import TableCell from './TableCell';

const cn = cnCreate('mfui-beta-table');

const props = {
    className: 'custom-class',
    fixColumn: false,
    minCellSize: 'small',
} as ITable;

describe('<Table />', () => {
    it('should render with default props', () => {
        const wrapper = shallow(
            <Table>
                <TableRow head>
                    <TableCell>Заголовок 1</TableCell>
                    <TableCell>Заголовок 2</TableCell>
                </TableRow>
                <TableRow>
                    <TableCell>Текст 1</TableCell>
                    <TableCell>Текст 2</TableCell>
                </TableRow>
            </Table>
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render with props', () => {
        const wrapper = shallow(
            <Table {...props}>
                <TableRow head>
                    <TableCell>Заголовок 1</TableCell>
                    <TableCell>Заголовок 2</TableCell>
                </TableRow>
                <TableRow>
                    <TableCell>Текст 1</TableCell>
                    <TableCell>Текст 2</TableCell>
                </TableRow>
            </Table>
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render cells as converted string with nbsp', () => {
        const wrapper = shallow(
            <Table {...props}>
                <TableRow head>
                    <TableCell>Заголовок&nbsp;1</TableCell>
                    <TableCell>Заголовок&nbsp;2</TableCell>
                </TableRow>
                <TableRow>
                    <TableCell>Текст&nbsp;1</TableCell>
                    <TableCell>Текст&nbsp;2</TableCell>
                </TableRow>
            </Table>
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render cells as react node', () => {
        const wrapper = shallow(
            <Table {...props}>
                <TableRow head>
                    <TableCell>
                        <span>Заголовок 1</span>
                    </TableCell>
                    <TableCell>
                        <span>Заголовок 2</span>
                    </TableCell>
                </TableRow>
                <TableRow>
                    <TableCell>
                        <span>Текст 2</span>
                    </TableCell>
                    <TableCell>
                        <span>Текст 2</span>
                    </TableCell>
                </TableRow>
            </Table>
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render with touch mod', () => {
        const onTouchStart = window.ontouchstart;
        window.ontouchstart = jest.fn();

        const wrapper = mount(
            <Table>
                <TableRow head>
                    <TableCell>Заголовок 1</TableCell>
                    <TableCell>Заголовок 2</TableCell>
                </TableRow>
                <TableRow>
                    <TableCell>Текст 1</TableCell>
                    <TableCell>Текст 2</TableCell>
                </TableRow>
            </Table>
        );

        expect(wrapper.children().hasClass(`${cn()}_touch`)).toBeTruthy();

        window.ontouchstart = onTouchStart;
    });
});
