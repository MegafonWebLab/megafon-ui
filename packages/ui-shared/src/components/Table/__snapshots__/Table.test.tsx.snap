// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should render cells as converted string with nbsp 1`] = `
<div
  className="mfui-beta-table mfui-beta-table_min-cell-size_small custom-class"
>
  <div
    className="mfui-beta-table__scroll"
    onScroll={[Function]}
  >
    <table
      className="mfui-beta-table__table"
    >
      <tbody>
        <tr
          className="mfui-beta-table__row mfui-beta-table__row_head"
          key=".0"
        >
          <td
            className="mfui-beta-table__cell"
            key=".0"
          >
            Заголовок 1
          </td>
          <td
            className="mfui-beta-table__cell"
            key=".1"
          >
            Заголовок 2
          </td>
        </tr>
        <tr
          className="mfui-beta-table__row"
          key=".1"
        >
          <td
            className="mfui-beta-table__cell"
            key=".0"
          >
            Текст 1
          </td>
          <td
            className="mfui-beta-table__cell"
            key=".1"
          >
            Текст 2
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`<Table /> should render cells as react node 1`] = `
<div
  className="mfui-beta-table mfui-beta-table_min-cell-size_small custom-class"
>
  <div
    className="mfui-beta-table__scroll"
    onScroll={[Function]}
  >
    <table
      className="mfui-beta-table__table"
    >
      <tbody>
        <tr
          className="mfui-beta-table__row mfui-beta-table__row_head"
          key=".0"
        >
          <td
            className="mfui-beta-table__cell"
            key=".0"
          >
            <span>
              Заголовок 1
            </span>
          </td>
          <td
            className="mfui-beta-table__cell"
            key=".1"
          >
            <span>
              Заголовок 2
            </span>
          </td>
        </tr>
        <tr
          className="mfui-beta-table__row"
          key=".1"
        >
          <td
            className="mfui-beta-table__cell"
            key=".0"
          >
            <span>
              Текст 2
            </span>
          </td>
          <td
            className="mfui-beta-table__cell"
            key=".1"
          >
            <span>
              Текст 2
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`<Table /> should render with default props 1`] = `
<div
  className="mfui-beta-table mfui-beta-table_fix-column mfui-beta-table_min-cell-size_large"
>
  <div
    className="mfui-beta-table__scroll"
    onScroll={[Function]}
  >
    <table
      className="mfui-beta-table__table"
    >
      <tbody>
        <tr
          className="mfui-beta-table__row mfui-beta-table__row_head"
          key=".0"
        >
          <td
            className="mfui-beta-table__cell"
            key=".0"
          >
            Заголовок 1
          </td>
          <td
            className="mfui-beta-table__cell"
            key=".1"
          >
            Заголовок 2
          </td>
        </tr>
        <tr
          className="mfui-beta-table__row"
          key=".1"
        >
          <td
            className="mfui-beta-table__cell"
            key=".0"
          >
            Текст 1
          </td>
          <td
            className="mfui-beta-table__cell"
            key=".1"
          >
            Текст 2
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`<Table /> should render with props 1`] = `
<div
  className="mfui-beta-table mfui-beta-table_min-cell-size_small custom-class"
>
  <div
    className="mfui-beta-table__scroll"
    onScroll={[Function]}
  >
    <table
      className="mfui-beta-table__table"
    >
      <tbody>
        <tr
          className="mfui-beta-table__row mfui-beta-table__row_head"
          key=".0"
        >
          <td
            className="mfui-beta-table__cell"
            key=".0"
          >
            Заголовок 1
          </td>
          <td
            className="mfui-beta-table__cell"
            key=".1"
          >
            Заголовок 2
          </td>
        </tr>
        <tr
          className="mfui-beta-table__row"
          key=".1"
        >
          <td
            className="mfui-beta-table__cell"
            key=".0"
          >
            Текст 1
          </td>
          <td
            className="mfui-beta-table__cell"
            key=".1"
          >
            Текст 2
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
